HTML API: Fix invalid comment syntax cases

dmsnell opened the pull request:
## Status

 - [x] Satisfy linter requirements
 - [x] Create Trac ticket [#58007-trac](https://core.trac.wordpress.org/ticket/58007#ticket)
 - [x] Expand comments, docblocks, add version and Trac ticket refereces
 - [ ] Create backport in Gutenberg

## What

Adds support for a few invalid HTML comment forms.

 - comments created by means of a tag closer with an invalid tag name, e.g. `</3>`
 - comments closed with the invalid `--!>` closer (comments should be closed by `-->` but if the `!` appears it will also close it, in error).
 - empty tag name elements, which are technically skipped over and aren't comments, e.g. `</>`

## Why

We want to continue to find and handle edge cases in HTML so that the Tag Processor remains reliable in the face of unexpected inputs.

## How

Captures the patterns described in the HTML specification for parse errors relating to invalid comment construction. Since these are comments it's appropriate to proceed consuming input until the end, and there's no need to look for other tags or attributes in the process.

cc: @adamziel @gziolo

dmsnell committed:
HTML API: Support extra non-normative comment constructions

dmsnell committed:
Fix newly introduced and broken tests

dmsnell committed:
Fix implementation detail, move by one not two dashes

dmsnell committed:
Move pointer increment into loop condition to avoid infinite loops

This is just a safety precaution to mitigate an accidental removal of the
pointer increment. No actual problems motivated this change.

dmsnell committed:
Add references to Trac ticket, annotate tests

costdev commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1039,13 +1040,40 @@ private function parse_next_tag() {
-						return false;
+					$closer_at = $at + 4;
+
+					/*
+					 * Abruptly-closed empty comments are a sequence of dashes followed by `>`.
+					 */
```
```suggestion
					 // Abruptly-closed empty comments are a sequence of dashes followed by `>`.
```
This should use the single line comment format.

dmsnell replied:
inlined in ddcb554583

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1664,6 +1664,47 @@ public function data_next_tag_ignores_script_tag_contents() {
+	 *
+	 * See https://html.spec.whatwg.org/#parse-error-invalid-first-character-of-tag-name
+	 *
+	 * @ticket 58007
+	 * @dataProvider data_next_tag_ignores_invalid_first_character_of_tag_name_comments
```
```suggestion
	 * @ticket 58007
	 *
	 * @dataProvider data_next_tag_ignores_invalid_first_character_of_tag_name_comments
```
`@ticket` and `@dataProvider` should be separated by an empty line

dmsnell replied:
split in ddcb554583

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1664,6 +1664,47 @@ public function data_next_tag_ignores_script_tag_contents() {
+	 * @ticket 58007
+	 * @dataProvider data_next_tag_ignores_invalid_first_character_of_tag_name_comments
+	 *
+	 * @param string $html_with_markers HTML containing an invalid tag closer whose element before and element after contain the "start" and "end" CSS classes.
```
```suggestion
	 * @param string $html_with_markers HTML containing an invalid tag closer whose element before
                                          and element after contain the "start" and "end" CSS classes.
```
Let's break this line to limit the need for horizontal scrolling.

dmsnell replied:
broken in ddcb554583.

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1664,6 +1664,47 @@ public function data_next_tag_ignores_script_tag_contents() {
+
+	/**
+	 * Data provider.
+	 *
+	 * @ticket 58007
```
```suggestion
```
Data providers don't need the `@ticket` annotation

dmsnell replied:
removed in ddcb554583

costdev commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1039,13 +1040,40 @@ private function parse_next_tag() {
+							return false;
+						}
+
+						if ( '>' === $html[ $closer_at + 2 ] ) {
```
Is it possible that `$closer_at + 2` may not exist as an offset?

dmsnell replied:
yes it was, and thanks for noticing! I've added new tests and fixes in ae59e1a and 926d491fe3 which validate this particular case of unchecked indices

costdev replied:
Thanks @dmsnell! I love the dataset names! 😆

costdev reviewed (changes requested):
Thanks for the PR @dmsnell!

I've added some docs-related feedback and just one question regarding whether it's possible that an array offset may not exist.

dmsnell committed:
Linting issues

dmsnell committed:
Correct unchecked indexing, feedback

Props to @costdev for noting the unchecked indices.

dmsnell committed:
Catch another indexing operation

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1664,6 +1664,47 @@ public function data_next_tag_ignores_script_tag_contents() {
+	 * See https://html.spec.whatwg.org/#parse-error-invalid-first-character-of-tag-name
+	 *
+	 * @ticket 58007
+	 *
+	 * @dataProvider data_next_tag_ignores_invalid_first_character_of_tag_name_comments
```
```suggestion
	 *
	 * @covers WP_HTML_Tag_Processor::next_tag
	 *
	 * @dataProvider data_next_tag_ignores_invalid_first_character_of_tag_name_comments
```
See [this ticket](https://core.trac.wordpress.org/ticket/56070) as a reference for the test docblock order/spacing of annotations.

dmsnell replied:
spaced out in 8813327002

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1716,6 +1757,96 @@ public function data_next_tag_ignores_contents_of_rcdata_tag() {
+	/**
+	 * Ensures that the invalid comment closing syntax "--!>" properly closes a comment.
+	 *
+	 * @ticket 58007
+	 * @covers WP_HTML_Tag_Processor::next_tag
```
```suggestion
	 * @ticket 58007
	 *
	 * @covers WP_HTML_Tag_Processor::next_tag
```

dmsnell replied:
spaced out in 8813327002

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1716,6 +1757,96 @@ public function data_next_tag_ignores_contents_of_rcdata_tag() {
+	 *
+	 * @ticket 58007
+	 *
+	 * @covers WP_HTML_Tag_Processor::next_tag
+	 * @dataProvider data_html_with_unclosed_comments
```
```suggestion
	 * @covers WP_HTML_Tag_Processor::next_tag
	 *
	 * @dataProvider data_html_with_unclosed_comments
```

dmsnell replied:
spaced out in 8813327002

costdev commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1716,6 +1757,96 @@ public function data_next_tag_ignores_contents_of_rcdata_tag() {
+	 *
+	 * @ticket 58007
+	 *
+	 * @covers WP_HTML_Tag_Processor::next_tag
+	 * @dataProvider data_abruptly_closed_empty_comments
```
```suggestion
	 * @covers WP_HTML_Tag_Processor::next_tag
	 *
	 * @dataProvider data_abruptly_closed_empty_comments
```

dmsnell replied:
spaced out in 8813327002

dmsnell committed:
Format docblocks

dmsnell commented:
@costdev I think I've addressed all your comments now, thanks. must say I tried to get the docblock ordering and formatting right, but for some reason that's consistently hard to do

costdev reviewed (approved):
Thanks for the updates @dmsnell! Looks like there's just a PHPCS issue to resolve, but other than that this looks good to me 👍

dmsnell committed:
Linting issues

dmsnell commented:
So there were @costdev - sorry for missing that. I'll double-check once the tests finish again to make sure no remain.

ockham commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1118,6 +1154,22 @@ private function parse_next_tag() {
+			 * If a non-alpha starts the tag name in a tag closer it's a comment.
+			 * Find the first `>`, which closes the comment.
+			 *
+			 * See https://github.com/WordPress/wordpress-develop/pull/4256
```
Would it make sense to link to the spec as a somewhat more canonical source instead? While it's fair that this PR states the intent behind this change, I've been itching to see _why_ we can actually do this 😅 

How about e.g.

```suggestion
			 * See https://html.spec.whatwg.org/#parse-error-invalid-first-character-of-tag-name
```

(which is the reference we're also using [in the tests](https://github.com/WordPress/wordpress-develop/pull/4256/files#diff-f2dca46bdddb8b5f5c56051017e0d0ccd262b0401ba7bdb9db506187a40220b3R1672))

dmsnell replied:
good catch; this was probably me pasting the wrong thing and not noticing it

dmsnell replied:
updated in 663055b55f

ockham commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -1716,6 +1757,99 @@ public function data_next_tag_ignores_contents_of_rcdata_tag() {
+			'Comment with almost-closer inside'  => array( '<hr><!-- ---!><hr id=after>--!><hr id=final>' ),
+		);
+	}
+
```
Should we add test coverage for [`</>`](https://github.com/WordPress/wordpress-develop/pull/4256/files#diff-0c2d7bb49853cc0faafd5412a25ee90e8cbb4f9293e7c3bb308494aba7d08272R1133-R1142)?

dmsnell replied:
this is not technically an empty comment, but it is already covered in the `test_updating_attributes` function.

ockham replied:
Gotcha, thanks for the pointer!

dmsnell committed:
Update link referring to HTML spec instead of GitHub PR

dmsnell committed:
Avoid regression where comment opening is next-to-last character in document.

dmsnell committed:
Update docblock to mentionm abruptly-closed comment support.

dmsnell committed:
Update docblock to indicate support for empty tags

ockham reviewed (approved):
LGTM, thank you!

ockham commented:
Committed to Core `trunk` in [`r55667`](https://core.trac.wordpress.org/changeset/55667).

ockham commented:
Backported to Core's `6.2` branch in [`r55668`](https://core.trac.wordpress.org/changeset/55668).
